src/storage/src/hummock/iterator/merge_inner.rs

```diff
@@ -88,22 +90,122 @@ impl<'a, const DIRECTION: usize> MergeIteratorInner<'a, DIRECTION> {
 
         self.heap = self
             .unused_iters
-            .drain_filter(|i| i.is_valid())
-            .map(Node)
+            .drain_filter(|i| i.iter.is_valid())
             .collect();
     }
 }
 
 #[async_trait]
-impl<const DIRECTION: usize> HummockIterator for MergeIteratorInner<'_, DIRECTION> {
-    async fn next(&mut self) -> HummockResult<()> {
+/// The behaviour of `next` of order aware merge iterator is different from the normal one, so we
+/// extract this trait.
+trait MergeIteratorNext<'a> {
+    async fn next_inner(&mut self) -> HummockResult<()>;
+}
+
+/// An order aware merge iterator. `usize` as the `extra_info` is used to define the order of
+/// iterators.
+pub type OrderedMergeIteratorInner<'a, D> = MergeIteratorInner<'a, D, usize>;
+
+impl<'a, D: HummockIteratorDirection> OrderedMergeIteratorInner<'a, D> {
+    #[allow(dead_code)]
+    pub fn new(
+        iterators: impl IntoIterator<Item = BoxedHummockIterator<'a, D>>,
+        stats: Arc<StateStoreMetrics>,
+    ) -> Self {
+        Self {
+            unused_iters: iterators
+                .into_iter()
+                .enumerate()
+                .map(|(i, iter)| Node {
+                    iter,
+                    extra_info: i,
+                })
+                .collect(),
+            heap: BinaryHeap::new(),
+            stats,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a, D: HummockIteratorDirection> MergeIteratorNext<'a> for OrderedMergeIteratorInner<'a, D> {
+    async fn next_inner(&mut self) -> HummockResult<()> {
+        let top_node = self.heap.pop().expect("no inner iter");
+        let mut popped_nodes = vec![];
+
+        // Take all nodes with the same current key as the top_node out of the heap.
```

Comment:
Why do we need to take all nodes out? I believe we can simply pop they by combination of key + extra data.

Reply:
Here we just pop the nodes with the same current key as the top node, not all nodes.

Reply:
Yes, but why do we need to do this? I think the heap by default supports handling equal keys?

Reply:
The key stored in the heap is (iter_current_key, order_index), so that if we have multiple same iter_current_key, the iter with the smallest order_index appears at the top. 

We need to take out all nodes on the heap top with the same iter_current_key only, and in the view of heap, those keys are all different.

Reply:
Can we emit the same key (with same epoch) multiple times to the upper UserKeyIterator? Then we can write in the same way as before.

Reply:
But I guess there will be performance regression 😛 not sure

Reply:
> Can we emit the same key (with same epoch) multiple times to the upper UserKeyIterator

I've come up with a better idea. We can move this logic from upper iterator to the merge iterator itself. 

Then the logic will be like
```
let mut heap_top = heap.peek_mut();
let top_key = heap_top.iter.key();
heap_top.iter.next()
drop(heap_top);
while ... {
    let mut heap_top =  heap.peek_mut();
    if heap_top.iter.key() != top_key { break; }
    heap_top.iter.next();
    drop(heap_top);
}
```

Then we can avoid taking out the top node every time.

Reply:
`peek_mut` doesn't take out the top node. The node stays at its original place in heap. When `PeekMut` object is dropped, `BinaryHeap` will drift down the node. So I don't see how this optimization could benefit.

Also `MergeIterator` should emit all keys of its underlying iterators. Otherwise data will be lost when doing compaction. OrderedMergeIterator might be an exception though, because watermark is calculated on epoch level instead of epoch + order.

Reply:
> So I don't see how this optimization could benefit.

The current `OrderedMergeIterator` will by default take out the top node so that the second topmost node can be visible and only then we check whether we have advanced all the iterators with the same current key. This is the reason why I separate the `next` of the ordered and unordered version, since taking out the top node and re-push it back to the heap will increase the cost, while for the unordered version it is unnecessary to bear this cost. Now with this optimization, for both ordered and unordered version we don't have to take out the top node, and then they can share the same logic, and the performance of the unordered version will be almost the same.

Reply:
> Now with this optimization, for both ordered and unordered version we don't have to take out the top node, and then they can share the same logic, and the performance of the unordered version will be almost the same.

Okay, I totally agree on this point. But consider the correctness issue:

> Also MergeIterator should emit all keys of its underlying iterators. Otherwise data will be lost when doing compaction.

The pseudo code above directly compared key -- `if heap_top.iter.key() != top_key { break; }`, which means that for a given key, only its latest epoch will be emitted.

This will cause compactor to lose data. Compactor is using MergeIterator, and it requires keys within watermark..latest are all retiained.

Reply:
> The pseudo code above directly compared key

Here I mean the full key (user_key, epoch).

Reply:
... so if we want to unify the logic, I'd prefer to emit all keys in both ordered / unordered case. But I'm not sure if this would affect benchmark result. I'm okay with the current implementation, and I just want to follow up on this conversation I randomly recalled just now 🤣

Reply:
> Here I mean the full key (user_key, epoch).

Then there's no problem 🤓